### DISCUSSION TITLE
Your prisma schema has changed, please restart Keystone

### DISCUSSION THREAD

**Discussion by alfreddohnani** [keystonejs/keystone#7454] [Questions] [answered]
1. Initialise a new keystone6 project
2. Create some schemas
3. Run yarn dev

Keystone performs initial migration but fails to start. The output says "your prisma schema has changed...", yet there has been no new changes to any schema in keystone

<img width="422" alt="Screenshot 2022-04-04 at 8 21 08 PM" src="https://user-images.githubusercontent.com/54582881/161625423-bc369527-7ddc-466f-8704-588b7142c2c9.png">

node version: v14.19.0
macOs: 11.6.5
(PostgreSQL) 14.0

**Comment by dcousens** [comment]
@mesteralfreddohnani 
> Initialise a new keystone6 project

Could you expand on what you are doing at this step?

**Comment by alfreddohnani** [comment]
> @mesteralfreddohnani
> 
> > Initialise a new keystone6 project
> 
> Could you expand on what you are doing at this step?

I scaffolded the project with `yarn create keystone-app crud-service`. Here is a repo to reproduce the issue: [https://github.com/mesteralfreddohnani/crud-service](url)

**Comment by alfreddohnani** [comment, marked as answer]
<img width="578" alt="Screenshot 2022-04-05 at 6 56 51 AM" src="https://user-images.githubusercontent.com/54582881/161696564-442ca91a-d51c-4433-b1ce-0577c01f9939.png">
Sorry, my bad. I think I know what the issue is. On `line 23` of the `Organisation` schema, the intention was to add 14 days to the current date before the data is saved. But it appears it thinks the default value changes every time. How do I best go about the above intention?

**Reply by dcousens** [reply]
Hi @mesteralfreddohnani,

Please refer to our [`default-values` example](https://github.com/keystonejs/keystone/tree/3bb1a53434b86e8a6294cff01a8699c36dd5df5a/examples/default-values) for how you can do dynamic default values using a `validateInput` hook.

**Comment by PhilBookst** [comment]
Hey I'm getting the same error after upgrading to the newest release, from the version before 

It's building the app correctly but after the point `Admin UI ready` it fails with the message 
`Your prisma schema has changed, please restart Keystone`

Any idea what could be causing this?

**Reply by PhilBookst** [reply]
If I build my project with yarn build and then yarn start the keystone app crashes recursively with the following error: 

```
TypeError: Cannot read properties of undefined (reading 'match')
    at Module.wt (_app-9b074f160a67106d.js:256:5126)
    at _app-9b074f160a67106d.js:1:153268
    at Array.forEach (<anonymous>)
    at _app-9b074f160a67106d.js:1:152394
    at Array.forEach (<anonymous>)
    at _app-9b074f160a67106d.js:1:152257
    at Object.To [as useMemo] (framework-a3035d159901c12f.js:1:64149)
    at t.useMemo (framework-a3035d159901c12f.js:1:125293)
    at k (_app-9b074f160a67106d.js:1:152100)
    at C (_app-9b074f160a67106d.js:1:154757)
```

**Reply by tranlehaiquan** [reply]
Hi @PhilBookst ,
I think you got the same problem as me when you want to use SQLite for local and Postgres for production. **You can resolve this by using only SQLite or Postgres**. And now the reason why:
In my .env I set NODE_ENV to production and log it in keystone.ts, but as you see it so production at the first time and development in the next time -> it will think you are changing from SQLite to Postgres and show the message `Your Prisma schema has changed, please restart Keystone`.

![Screen Shot 2022-04-19 at 01 58 38](https://user-images.githubusercontent.com/17347993/163861125-a60760be-1d37-4257-af04-99e29a7be496.png)

![Screen Shot 2022-04-19 at 01 58 24](https://user-images.githubusercontent.com/17347993/163861187-3ba8c7af-b16c-4801-a9e7-e2714c5ec8a9.png)

**Reply by PhilBookst** [reply]
Hmm no I'm using postgres both in development and production. Also my NODE_ENV is not set to production. 
I really have no clue on how to debug this

**Reply by PhilBookst** [reply]
Ok I found the issue:

in my schema.ts I had the following logic:
```ts
import * as List from "./schema"

// Schema lists aggregation point
export const lists = {
   ...List
}
```

my logs then showed 
```js
Lists {
  User: [Getter],
  Role: [Getter],
  Post: [Getter],
  Category: [Getter],
  Subcategory: [Getter],
  Field: [Getter],
  Keyword: [Getter],
  Faq: [Getter],
  FaqItem: [Getter],
  Picture: [Getter],
  FlashCard: [Getter],
  Diagnose: [Getter],
  BodyPart: [Getter],
  Symptom: [Getter],
  MedicalCondition: [Getter],
  DifferentialDiagnosis: [Getter],
  MedicalQuestionList: [Getter],
  MedicalQuestion: [Getter],
  MedicalAnswer: [Getter]
}
⭐️ Dev Server Starting on http://localhost:3000
⭐️ GraphQL API Starting on http://localhost:3000/api/graphql
✨ Generating GraphQL and Prisma schemas
✨ Your database is up to date, no migrations need to be created or applied
✨ Connecting to the database
✨ Creating server
✅ GraphQL API ready
✨ Generating Admin UI code
✨ Preparing Admin UI app
event - compiled client and server successfully in 7.8s (1728 modules)
✅ Admin UI ready
wait  - compiling /api/__keystone_api_build...
event - compiled client and server successfully in 874 ms (1839 modules)
Lists Object [Module] {
  BodyPart: [Getter],
  Category: [Getter],
  Diagnose: [Getter],
  DifferentialDiagnosis: [Getter],
  Faq: [Getter],
  FaqItem: [Getter],
  Field: [Getter],
  FlashCard: [Getter],
  Keyword: [Getter],
  MedicalCondition: [Getter],
  Picture: [Getter],
  Post: [Getter],
  Role: [Getter],
  Subcategory: [Getter],
  Symptom: [Getter],
  User: [Getter],
  MedicalAnswer: [Getter],
  MedicalQuestion: [Getter],
  MedicalQuestionList: [Getter]
}
```

so the order was messed up which lead to the "schema change"

**Reply by dcousens** [reply]
Thanks for following up with your resolution @PhilBookst

**Reply by pkuwyn** [reply]
`import * as lists from "./schema";`
using this import statement ,The order of lists changes randomly after update to the march 25 release.

While using this statement, the order remains unchanged:
`import {List1, List2} from "./schema";
const lists= {List1, List2};
`
However, this hardcode way of importing List is not a good practice, why this happened, any idea?

**Reply by KRIMOD** [reply]
i stumbled through the same issue, hardcoding the import as @pkuwyn explained solved it. 
Still don't know why it changes the order of the imports, may be JS object distructuring ?

**Reply by acburdine** [reply]
I ran into this issue after updating my version of Next.js to 12.1.6. 

My setup was like this:

```js
// keystone.ts
import * as lists from "./models";

export default config({
  // ...
  lists,
});

// models/index.ts
export { default as ModelB } from "./ModelB";
export { default as ModelA } from "./ModelA";
```

To fix the issue, i had to alphabetize the exports in `models/index.ts`:
```js
// models/index.ts
export { default as ModelA } from "./ModelA";
export { default as ModelB } from "./ModelB";
```

my theory is that the initial import of keystone.ts directly by Keystone Core preserves the original order of list elements. However, the `/api/__keystone_api_build` route is compiled by the Next.js admin, so there could be something in the Nextjs compiler that is re-ordering the list elements and alphabetizing them, causing Keystone to detect a change in the order of list items

**Reply by selcukok** [reply]
@acburdine ![Screenshot_2022-06-10-00-50-34-94_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/106354809/172951687-e4584188-048f-4296-acd1-48c82f2b8f38.jpg)

**Comment by selcukok** [comment]
![Screenshot_2022-06-09-23-56-28-89_e307a3f9df9f380ebaf106e1dc980bb6.jpg](https://user-images.githubusercontent.com/106354809/172951750-3ab566dd-7775-4aa4-9cfa-9a79c05b8973.jpg)
